File: backend/scripts/fetchProfilePosts.js
#!/usr/bin/env node

/**
 * @fileoverview Script to fetch one post/reel for each Instagram profile
 * @author Backend Team
 */

const fs = require('fs');
const path = require('path');
const { validateInstagramUsername } = require('../utils/instagram/profileResolver');

/**
 * Read profile links from text file
 * @param {string} filePath - Path to profile links file
 * @returns {Array<string>} Array of profile URLs
 */
function readProfileLinks(filePath) {
    try {
        const content = fs.readFileSync(filePath, 'utf-8');
        const lines = content.split('\n');
        
        // Skip header and filter valid URLs
        return lines
            .slice(1)
            .map(line => line.trim())
            .filter(line => line.startsWith('http'));
    } catch (error) {
        console.error('Error reading profile links:', error);
        return [];
    }
}

/**
 * Extract username from Instagram URL
 * @param {string} profileUrl - Instagram profile URL
 * @returns {string} Username
 */
function extractUsername(profileUrl) {
    return profileUrl
        .replace('https://www.instagram.com/', '')
        .replace('/', '')
        .trim();
}

/**
 * Generate sample post URLs for profiles
 * This is a placeholder - in production, you'd fetch actual posts
 * @param {Array<string>} profileUrls - Array of profile URLs
 * @returns {Array<Object>} Array of profile-post mappings
 */
function generateProfilePostMappings(profileUrls) {
    const mappings = [];
    
    // Sample post IDs for demonstration
    const samplePostIds = [
        'C1234567890A', 'C2345678901B', 'C3456789012C', 
        'C4567890123D', 'C5678901234E', 'C6789012345F'
    ];
    
    profileUrls.forEach((profileUrl, index) => {
        const username = extractUsername(profileUrl);
        const validation = validateInstagramUsername(username);
        
        if (validation.valid) {
            // For demonstration, alternate between posts and reels
            const isReel = index % 2 === 0;
            const postType = isReel ? 'reel' : 'p';
            const postId = samplePostIds[index % samplePostIds.length] + index;
            
            mappings.push({
                profile_url: profileUrl,
                username: validation.cleanUsername,
                post_url: `https://www.instagram.com/${postType}/${postId}/`,
                post_type: isReel ? 'reel' : 'post',
                status: 'sample'
            });
        } else {
            mappings.push({
                profile_url: profileUrl,
                username: username,
                post_url: '',
                post_type: '',
                status: 'invalid_username'
            });
        }
    });
    
    return mappings;
}

/**
 * Save mappings to CSV file
 * @param {Array<Object>} mappings - Profile-post mappings
 * @param {string} outputPath - Output CSV file path
 */
function saveToCSV(mappings, outputPath) {
    const headers = ['profile_url', 'username', 'post_url', 'post_type', 'status'];
    const csvContent = [
        headers.join(','),
        ...mappings.map(row => 
            headers.map(header => `"${row[header] || ''}"`).join(',')
        )
    ].join('\n');
    
    fs.writeFileSync(outputPath, csvContent);
    console.log(`✓ Saved ${mappings.length} mappings to ${outputPath}`);
}

/**
 * Main function
 */
async function main() {
    console.log('=' .repeat(60));
    console.log('Instagram Profile-Post Mapper');
    console.log('=' .repeat(60));
    
    const profileLinksPath = path.join(process.cwd(), 'profile_link.txt');
    const outputPath = path.join(process.cwd(), 'profile_posts.csv');
    
    // Read profile links
    console.log('\n1. Reading profile links...');
    const profileUrls = readProfileLinks(profileLinksPath);
    console.log(`   Found ${profileUrls.length} profiles`);
    
    if (profileUrls.length === 0) {
        console.error('No profiles found in profile_link.txt');
        process.exit(1);
    }
    
    // Generate mappings
    console.log('\n2. Generating profile-post mappings...');
    const mappings = generateProfilePostMappings(profileUrls);
    
    // Show sample
    console.log('\n3. Sample mappings:');
    mappings.slice(0, 5).forEach(mapping => {
        console.log(`   ${mapping.username} -> ${mapping.post_type || 'N/A'}`);
    });
    
    // Save to CSV
    console.log('\n4. Saving to CSV...');
    saveToCSV(mappings, outputPath);
    
    console.log('\n✓ Complete!');
    console.log('\nNote: This script generates sample post URLs for demonstration.');
    console.log('To fetch actual posts, you would need to:');
    console.log('1. Use Instagram\'s Graph API with proper authentication');
    console.log('2. Use the Chrome extension to gather data while browsing');
    console.log('3. Manually collect post URLs from each profile');
}

// Run if called directly
if (require.main === module) {
    main().catch(error => {
        console.error('Error:', error);
        process.exit(1);
    });
}

module.exports = {
    readProfileLinks,
    extractUsername,
    generateProfilePostMappings,
    saveToCSV
};

File: fetch_profile_posts.py
#!/usr/bin/env python3

import csv
import time
import json
import re
import requests
from urllib.parse import quote

def fetch_instagram_post(username):
    """
    Fetch the first post/reel from an Instagram profile.
    This uses Instagram's public web interface.
    """
    try:
        # Clean username
        username = username.replace('https://www.instagram.com/', '').strip('/')
        
        # Try to fetch profile page
        headers = {
            'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36'
        }
        
        # Instagram's web interface URL
        url = f'https://www.instagram.com/{username}/'
        
        response = requests.get(url, headers=headers, timeout=10)
        
        if response.status_code == 200:
            # Look for post URLs in the HTML
            # Instagram posts have pattern /p/XXXXX/ or /reel/XXXXX/
            post_pattern = r'"/p/([^"]+)/"'
            reel_pattern = r'"/reel/([^"]+)/"'
            
            # Find posts
            posts = re.findall(post_pattern, response.text)
            reels = re.findall(reel_pattern, response.text)
            
            # Return the first post or reel found
            if posts:
                return f'https://www.instagram.com/p/{posts[0]}/'
            elif reels:
                return f'https://www.instagram.com/reel/{reels[0]}/'
            else:
                # Try to find in script tags
                script_pattern = r'"shortcode":"([^"]+)"'
                shortcodes = re.findall(script_pattern, response.text)
                if shortcodes:
                    return f'https://www.instagram.com/p/{shortcodes[0]}/'
        
        return None
        
    except Exception as e:
        print(f"Error fetching {username}: {str(e)}")
        return None

def process_profiles(input_file='profile_link.txt', output_csv='profile_posts.csv'):
    """
    Process profile links and fetch one post for each profile.
    """
    # Read profile links
    profiles = []
    with open(input_file, 'r', encoding='utf-8') as file:
        lines = file.readlines()
        for line in lines[1:]:  # Skip header
            url = line.strip()
            if url and url.startswith('http'):
                profiles.append(url)
    
    print(f"Processing {len(profiles)} profiles...")
    print("Note: This will take some time to avoid rate limiting.")
    print("Press Ctrl+C to stop and save progress.\n")
    
    results = []
    processed = 0
    
    try:
        for profile_url in profiles:
            username = profile_url.replace('https://www.instagram.com/', '').strip('/')
            print(f"[{processed + 1}/{len(profiles)}] Processing @{username}...", end=' ')
            
            # Fetch a post from this profile
            post_url = fetch_instagram_post(username)
            
            if post_url:
                print(f"✓ Found post")
                results.append({
                    'profile_url': profile_url,
                    'username': username,
                    'post_url': post_url
                })
            else:
                print(f"✗ No posts found")
                results.append({
                    'profile_url': profile_url,
                    'username': username,
                    'post_url': 'No posts found'
                })
            
            processed += 1
            
            # Rate limiting - wait between requests
            if processed < len(profiles):
                time.sleep(2)  # Wait 2 seconds between requests
            
            # Save progress every 10 profiles
            if processed % 10 == 0:
                save_results(results, output_csv)
                print(f"Progress saved: {processed}/{len(profiles)}")
    
    except KeyboardInterrupt:
        print("\n\nInterrupted by user. Saving progress...")
    
    # Save final results
    save_results(results, output_csv)
    print(f"\nProcessed {processed} profiles")
    print(f"Results saved to: {output_csv}")
    
    return results

def save_results(results, output_csv):
    """
    Save results to CSV file.
    """
    with open(output_csv, 'w', newline='', encoding='utf-8') as file:
        writer = csv.DictWriter(file, fieldnames=['profile_url', 'username', 'post_url'])
        writer.writeheader()
        writer.writerows(results)

def create_sample_data():
    """
    Create sample data for demonstration.
    Since we can't actually fetch from Instagram without authentication,
    we'll create a sample dataset.
    """
    print("Creating sample profile_posts.csv with demonstration data...")
    
    # Read profile links
    profiles = []
    with open('profile_link.txt', 'r', encoding='utf-8') as file:
        lines = file.readlines()
        for line in lines[1:]:  # Skip header
            url = line.strip()
            if url and url.startswith('http'):
                profiles.append(url)
    
    # Create sample post URLs
    results = []
    post_types = ['p', 'reel']
    sample_ids = [
        'C1234567890A', 'C2345678901B', 'C3456789012C', 'C4567890123D',
        'C5678901234E', 'C6789012345F', 'C7890123456G', 'C8901234567H'
    ]
    
    for i, profile_url in enumerate(profiles[:50]):  # Process first 50 for demo
        username = profile_url.replace('https://www.instagram.com/', '').strip('/')
        post_type = post_types[i % 2]
        post_id = sample_ids[i % len(sample_ids)] + str(i)
        post_url = f'https://www.instagram.com/{post_type}/{post_id}/'
        
        results.append({
            'profile_url': profile_url,
            'username': username,
            'post_url': post_url
        })
    
    # Save to CSV
    with open('profile_posts.csv', 'w', newline='', encoding='utf-8') as file:
        writer = csv.DictWriter(file, fieldnames=['profile_url', 'username', 'post_url'])
        writer.writeheader()
        writer.writerows(results)
    
    print(f"Created sample data with {len(results)} profile-post pairs")
    print("Saved to: profile_posts.csv")
    
    return results

if __name__ == "__main__":
    import sys
    
    if '--sample' in sys.argv:
        # Create sample data for demonstration
        create_sample_data()
    else:
        print("=" * 60)
        print("Instagram Profile Post Fetcher")
        print("=" * 60)
        print("\nNOTE: Due to Instagram's anti-scraping measures,")
        print("fetching posts requires authentication or may be blocked.")
        print("\nOptions:")
        print("1. Run with --sample to create demonstration data")
        print("2. Use Instagram's API with proper authentication")
        print("3. Manually collect post URLs")
        print("\nTo create sample data: python fetch_profile_posts.py --sample")
        print("\nTo attempt fetching (may be blocked):")
        print("python fetch_profile_posts.py --fetch")
        
        if '--fetch' in sys.argv:
            process_profiles()